[Winlogbeat] Add configuration for registry file flush timeout

taylor-swanson opened the pull request:
## What does this PR do?

- Added a new option ("winlogbeat.registry_flush") for configuring the
timeout for writting registry entries to disk. The previously hard-coded
value of 5 seconds is now the default value.

## Why is it important?

The value is currently hard-coded and environments that have lots of event activity will cause unnecessary CPU usage due to excessive flushing/writing. By tuning this value, the flush interval can be adjusted to reduce CPU churn.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

- Add the option to the configuration file (`winlogbeat.yml`):
```
winlogbeat.registry_flush: 30s
```
Default is `5s`. Value should be a duration value (i.e., `5s`, `1m`)

## Related issues

- Closes #29001

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29053/detail/PR-29053/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29053/detail/PR-29053/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29053/detail/PR-29053/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-29053/detail/PR-29053/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2021-11-19T16:47:56.846Z&rangeTo=2021-11-19T17:07:56.846Z&transactionName=Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=unknown&latencyAggregationType=avg&traceId=7b2f9290c0748a781ecce216bb5f8cd3&transactionId=cc3cc13588bc090a)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats



* Reason: `Aborted from #2`


* Start Time: 2021-11-19T16:57:56.846+0000


* Duration: 3 min 16 sec


* Commit: 9bd85419c1e8303c43cb31c7f07ac9b76a2a647f




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
Pinging @elastic/security-external-integrations (Team:Security-External Integrations)

andrewkroh commented on winlogbeat/beater/winlogbeat.go:
```diff
@@ -110,7 +110,7 @@ func (eb *Winlogbeat) setup(b *beat.Beat) error {
 
 	var err error
-	eb.checkpoint, err = checkpoint.NewCheckpoint(config.RegistryFile, 10, 5*time.Second)
+	eb.checkpoint, err = checkpoint.NewCheckpoint(config.RegistryFile, 10, config.RegistryFlush)
```
In order to make increasing the value more effective at throttling the number of writes, can you disable the max number of updates parameter (like change the `10` to a `-1` and update the checkpoint package to handle that). Or completely remove that parameter from checkpoint.

andrewkroh reviewed (commented):
Thanks for adding this. 👍

andrewkroh reviewed (changes requested):
The code LGTM. The new configuration option needs to be documented in the Winlogbeat config docs and added to the winlogbeat.reference.yml. Can you please add those two things.

taylor-swanson commented:
> The code LGTM. The new configuration option needs to be documented in the Winlogbeat config docs and added to the winlogbeat.reference.yml. Can you please add those two things.

Strange, I had added the new config option to the winlogbeat.reference.yml files, but it got reverted somehow in one of my commits. I'll get that added back in, and I'll add to the config docs as well.

andrewkroh reviewed (approved):
LGTM
